src/ObjWriting/XModel/Export/XModelBinWriter.cpp

```diff
@@ -0,0 +1,484 @@
+        MATERIAL_COEFFS = 0xC835,
+        MATERIAL_GLOW = 0xFE0C,
+        MATERIAL_REFRACTIVE = 0x7E24,
+        MATERIAL_SPECULAR_COLOR = 0x317C,
+        MATERIAL_REFLECTIVE_COLOR = 0xE593,
+        MATERIAL_REFLECTIVE = 0x7D76,
+        MATERIAL_BLINN = 0x83C7,
+        MATERIAL_PHONG = 0x5CD2
+    };
+
+    void PrepareVertexMerger(const XModelCommon& xmodel)
+    {
+        m_vertex_merger = VertexMerger(xmodel.m_vertices.size());
+
+        auto vertexOffset = 0u;
+        for (const auto& vertex : xmodel.m_vertices)
+        {
+            XModelVertexBoneWeights weights{0, 0};
+
+            if (vertexOffset < xmodel.m_vertex_bone_weights.size())
+                weights = xmodel.m_vertex_bone_weights[vertexOffset];
+
+            m_vertex_merger.Add(VertexMergerPos{vertex.coordinates[0],
+                                                vertex.coordinates[1],
+                                                vertex.coordinates[2],
+                                                xmodel.m_bone_weight_data.weights.empty() ? nullptr : &xmodel.m_bone_weight_data.weights[weights.weightOffset],
+                                                weights.weightCount});
+
+            vertexOffset++;
+        }
+    }
+
+    template<typename T> void Write(const T& data)
+    {
+        m_writer.write(reinterpret_cast<const char*>(&data), sizeof(T));
+    }
+
+    void WriteNullTerminatedString(std::string string)
+    {
+        m_writer.write(string.c_str(), string.size() + 1);
+    }
+
+    void WriteAlignedString(const std::string& string)
+    {
+        static constexpr uint8_t PADDING[4] = {0u, 0u, 0u, 0u};
+        const auto paddingSize = (4u - (string.size() + 1u) % 4u) % 4u;
+
+        WriteNullTerminatedString(string);
+        m_writer.write(reinterpret_cast<const char*>(PADDING), paddingSize);
+    }
+
+    void WriteComment(const std::string& comment)
+    {
+        Write(XModelBinHash::COMMENT);
+        WriteAlignedString(comment);
+    }
+
+    void WriteInt16(const int16_t hash, const int16_t value)
+    {
+        Write(hash);
+        Write(value);
+    }
+
+    void WriteUInt16(const int16_t hash, const uint16_t value)
+    {
+        Write(hash);
+        Write(value);
+    }
+
+    int16_t ClampFloatToShort(const float value) const
+    {
+        return std::clamp(static_cast<int16_t>(std::numeric_limits<int16_t>::max() * std::clamp(value, -1.0f, 1.0f)),
+                          std::numeric_limits<int16_t>::min(),
+                          std::numeric_limits<int16_t>::max());
+    }
+
+    uint8_t ClampFloatToUByte(const float value) const
+    {
+        return std::clamp(static_cast<uint8_t>(std::numeric_limits<uint8_t>::max() * std::clamp(value, -1.0f, 1.0f)),
+                          std::numeric_limits<uint8_t>::min(),
+                          std::numeric_limits<uint8_t>::max());
+    }
+
+    void WriteHeader(int16_t version)
+    {
+        WriteComment("OpenAssetTools XMODEL_BIN File");
```

Comment:
Including branding is nice here :) I usually would make it include the version number as well for the following reason:

When somebody reports a bug related to a file being wrong, not being able to read it or something similar, it is then possible to see which version this file was generated with.

You can do that like this:
```cpp
#include "GitVersion.h"
// ...
WriteComment("OpenAssetTools " GIT_VERSION " XMODEL_BIN file");
```
The GitVersion.h file is generated by premake inside the build directory.

In C and C++, strings are concatinated statically when you place them next to each other without an operator.
So `"Hello" "World"` would result in `"HelloWorld"`.
That's how this works.